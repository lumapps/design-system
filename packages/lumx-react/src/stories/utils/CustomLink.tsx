/* eslint-disable jsx-a11y/anchor-has-content,jsx-a11y/anchor-is-valid */
import React from 'react';

/**
 * Example custom link that can be used in linkAs props.
 */
export const CustomLink: any = ({ ...props }) => <a href="#" data-custom-link {...props} />;
