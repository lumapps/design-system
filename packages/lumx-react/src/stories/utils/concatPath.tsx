export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
