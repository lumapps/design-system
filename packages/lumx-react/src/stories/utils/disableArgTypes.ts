export function disableArgTypes(argTypes: string[]) {
    return Object.fromEntries(argTypes.map((argType) => [argType, { table: { disable: true } }]));
}
