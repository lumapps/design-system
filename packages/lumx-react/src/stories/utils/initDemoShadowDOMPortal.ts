/** Init a demo shadow DOM to use as portal container */
export const initDemoShadowDOMPortal = () => {
    const div = document.createElement('div');
    document.body.appendChild(div);
    const container = div.attachShadow({ mode: 'open' });
    const style = document.createElement('style');
    style.innerText = `button { color: red; }`;
    container.appendChild(style);
    return { container, teardown: () => div.remove() };
};
