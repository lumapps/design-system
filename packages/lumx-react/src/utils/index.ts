/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
