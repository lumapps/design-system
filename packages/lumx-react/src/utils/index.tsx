export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
