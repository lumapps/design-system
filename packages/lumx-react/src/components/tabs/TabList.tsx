import { Alignment, Theme } from '@lumx/react';
import { CSS_PREFIX } from '@lumx/react/constants';
import { Comp, GenericProps, handleBasicClasses } from '@lumx/react/utils';
import { mergeRefs } from '@lumx/react/utils/mergeRefs';

import classNames from 'classnames';
import React, { forwardRef, ReactNode } from 'react';
import { useRovingTabIndex } from '../../hooks/useRovingTabIndex';

export enum TabListLayout {
    clustered = 'clustered',
    fixed = 'fixed',
}

/**
 * Defines the props of the component.
 */
export interface TabListProps extends GenericProps {
    /** ARIA label (purpose of the set of tabs). */
    ['aria-label']: string;
    /** Tab list. */
    children: ReactNode;
    /** Layout of the tabs in the list. */
    layout?: TabListLayout;
    /** Position of the tabs in the list (requires 'clustered' layout). */
    position?: Alignment;
    /** Theme adapting the component to light or dark background. */
    theme?: Theme;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'TabList';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = `${CSS_PREFIX}-tabs`;

/**
 * Component default props.
 */
const DEFAULT_PROPS: Partial<TabListProps> = {
    layout: TabListLayout.fixed,
    position: Alignment.left,
    theme: Theme.light,
};

/**
 * TabList component.
 *
 * Implements WAI-ARIA `tablist` role {@see https://www.w3.org/TR/wai-aria-practices-1.1/examples/tabs/tabs-1/tabs.html#rps_label}
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const TabList: Comp<TabListProps, HTMLDivElement> = forwardRef((props, ref) => {
    const { 'aria-label': ariaLabel, children, className, layout, position, theme, ...forwardedProps } = props;
    const tabListRef = React.useRef(null);
    useRovingTabIndex({
        parentRef: tabListRef,
        elementSelector: '[role="tab"]',
        keepTabIndex: false,
        extraDependencies: [children],
    });

    return (
        <div
            ref={mergeRefs(ref, tabListRef)}
            {...forwardedProps}
            className={classNames(className, handleBasicClasses({ prefix: CLASSNAME, layout, position, theme }))}
        >
            <div className={`${CLASSNAME}__links`} role="tablist" aria-label={ariaLabel}>
                {children}
            </div>
        </div>
    );
});
TabList.displayName = COMPONENT_NAME;
TabList.className = CLASSNAME;
TabList.defaultProps = DEFAULT_PROPS;
