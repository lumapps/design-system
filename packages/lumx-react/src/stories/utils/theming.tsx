import React from 'react';
import { Predicate } from '@lumx/react/utils/type';
import { FlexBox } from '@lumx/react';

type Property = string | { value: string };
interface PropertyTree {
    [key: string]: PropertyTree | Property;
}
type Entry = [key: string, value: string];
type Variables = Record<string, string>;

function asValue(prop?: PropertyTree | Property): string | undefined {
    if (typeof prop === 'string') {
        return prop;
    }
    if (typeof prop === 'object' && 'value' in prop) {
        return prop.value as string;
    }
    return undefined;
}

/** Convert property tree into a list of args entries ([key, value]) */
function* toArgEntries(props: PropertyTree | Property, parent = ''): Iterable<Entry> {
    const value = asValue(props);
    if (value) {
        yield [parent, value];
    } else {
        for (const [key, prop] of Object.entries(props)) {
            const newParent = parent ? `${parent}-${key}` : key;
            yield* toArgEntries(prop, newParent);
        }
    }
}
function* parseVariableEntries(cssString: string, strict = false) {
    for (const line of cssString.split(/(?<=;)/)) {
        if (!line.trim()) continue;
        const match = line.match(/^\s*(--lumx[^:]+)\s*:\s*([^;]+);/m);
        if (!match) {
            if (strict) throw new Error(`Could not parse \`${line}\``);
            else continue;
        }
        const [, variableName, variableValue] = match;
        yield [variableName, variableValue];
    }
}

function VariableEditor({
    variable: [key, value],
    setVariable,
    defaultStyle,
}: {
    variable: Entry;
    setVariable: (newVariable: Entry) => void;
    defaultStyle: Variables;
}) {
    const onChange = React.useCallback<React.ChangeEventHandler<HTMLInputElement>>(
        (evt) => {
            setVariable([evt.target.name, evt.target.value]);
        },
        [setVariable],
    );

    return (
        <FlexBox as="label" orientation="horizontal" hAlign="center" style={{ whiteSpace: 'nowrap' }}>
            <span style={{ flexShrink: 0 }}>{key}</span>:
            <input
                name={key}
                onChange={onChange}
                style={{ border: 0, padding: 8, fieldSizing: 'content', display: 'inline' } as any}
                value={defaultStyle[key] === value ? '' : value}
                placeholder={value}
            />
            ;
        </FlexBox>
    );
}

function VariablesEditor({
    style,
    defaultStyle,
    setStyle,
}: {
    style: Variables;
    defaultStyle: Variables;
    setStyle: (fn: (newStyle: Variables) => Variables) => void;
}) {
    const setVariable = React.useCallback(
        ([key, value]: Entry) => {
            setStyle((prevStyle) => ({ ...prevStyle, [key]: value }));
        },
        [setStyle],
    );
    return (
        <details style={{ fontFamily: 'monospace' }} className="lumx-spacing-margin-vertical-huge">
            <summary className="lumx-typography-subtitle2">Edit variables</summary>
            {Object.entries(style).map((variable) => (
                <VariableEditor
                    key={variable[0]}
                    defaultStyle={defaultStyle}
                    variable={variable}
                    setVariable={setVariable}
                />
            ))}
        </details>
    );
}

const FILTER_PROPS = ['theme', 'emphasis', 'size'];

export function withTheming<P extends PropertyTree>({
    properties,
    values,
}: {
    properties?: P;
    values?: string | Partial<P>;
}) {
    const initialStyle = values
        ? Object.fromEntries(
              typeof values === 'string'
                  ? parseVariableEntries(values)
                  : toArgEntries(values as PropertyTree, '--lumx'),
          )
        : [];
    const defaultStyle = Object.fromEntries(properties ? Array.from(toArgEntries(properties, '--lumx')) : []);

    const varNames = new Set(Object.keys(initialStyle).concat(Object.keys(defaultStyle)));

    function filteredStyles(args: any) {
        // Create predicate
        const filterArgs = Object.entries(args)
            .map(([argName, argValue]) => {
                if (typeof argValue === 'string' && FILTER_PROPS.includes(argName)) {
                    return (varName: string) =>
                        // Has the given arg with value
                        varName.match(new RegExp(`${argName}-${argValue}`)) ||
                        (argName === 'emphasis' && argValue === 'medium' && varName.match(/emphasis-selected/)) ||
                        // Or does not have the arg at all
                        !varName.match(new RegExp(`${argName}-\\w+`));
                }
                return false;
            })
            .filter(Boolean) as Array<Predicate<string>>;
        const styles = Object.fromEntries(
            Array.from(varNames.values())
                .map((varName) => {
                    if (!filterArgs.every((filter) => filter(varName))) return null as any;
                    return [varName, initialStyle[varName] || defaultStyle[varName]];
                })
                .filter(Boolean),
        );
        return () => styles;
    }

    return (Story: any, ctx: any) => {
        const [style, setStyle] = React.useState<Variables>(filteredStyles(ctx.args));

        return (
            <>
                <Story args={{ ...ctx.args, style: { ...ctx.args.style, ...style } }} />
                <VariablesEditor style={style} defaultStyle={defaultStyle} setStyle={setStyle} />
            </>
        );
    };
}
