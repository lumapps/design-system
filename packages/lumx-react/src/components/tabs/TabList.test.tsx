import { Tab, Alignment } from '@lumx/react';

import { Wrapper, commonTestsSuite } from '@lumx/react/testing/utils';
import { getBasicClass } from '@lumx/react/utils';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';
import React, { ReactElement } from 'react';
import { build, oneOf } from 'test-data-bot';
import { TabList, TabListLayout, TabListProps } from './TabList';
import { setupTabProviderMocks } from './test.mocks';

const CLASSNAME = TabList.className as string;

// Mock useTabProviderContext.
jest.mock('./state', () => {
    const state = jest.requireActual('./state');
    return { ...state, useTabProviderContext: jest.fn(), useTabProviderContextState: jest.fn() };
});

type SetupProps = Partial<TabListProps>;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = ({ ...propsOverride }: SetupProps = {}, shallowRendering = true) => {
    const tabs = [<Tab key={0} label="Tab 0" />, <Tab key={1} label="Tab 1" />];
    const props: TabListProps = {
        children: tabs,
        'aria-label': 'Tab list',
        ...propsOverride,
    };
    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;
    const wrapper = renderer(<TabList {...props} />);

    return { props, wrapper };
};

describe(`<${TabList.displayName}>`, () => {
    beforeEach(() => {
        setupTabProviderMocks();
    });

    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        it('should render correctly', () => {
            const { wrapper } = setup();
            expect(wrapper).toMatchSnapshot();

            expect(wrapper).toExist();
            expect(wrapper).toHaveClassName(CLASSNAME);

            expect(wrapper.childAt(0)).toExist();
            expect(wrapper.childAt(0)).toHaveClassName('lumx-tabs__links');
        });
    });

    // 2. Test defaultProps value and important props custom values.
    describe('Props', () => {
        it('should use the given props', () => {
            const modifiedPropsBuilder: () => SetupProps = build('props').fields({
                layout: TabListLayout.clustered,
                position: oneOf(Alignment.center, Alignment.right),
            });

            const modifiedProps: SetupProps = modifiedPropsBuilder();

            const { wrapper } = setup({ ...modifiedProps });

            Object.keys(modifiedProps).forEach((prop) => {
                expect(wrapper).toHaveClassName(
                    getBasicClass({ prefix: CLASSNAME, type: prop, value: modifiedProps[prop] }),
                );
            });
        });
    });

    // Common tests suite.
    commonTestsSuite(setup, { className: 'wrapper', prop: 'wrapper' }, { className: CLASSNAME });
});
