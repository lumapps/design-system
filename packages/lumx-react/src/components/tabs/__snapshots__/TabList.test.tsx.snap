// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TabList> Snapshots and structure should render correctly 1`] = `
<div
  className="lumx-tabs lumx-tabs--layout-fixed lumx-tabs--position-left lumx-tabs--theme-light"
>
  <div
    aria-label="Tab list"
    className="lumx-tabs__links"
    role="tablist"
  >
    <Tab
      key="0"
      label="Tab 0"
    />
    <Tab
      key="1"
      label="Tab 1"
    />
  </div>
</div>
`;
